sound/soc/sof/topology.c

```diff
@@ -952,10 +934,35 @@ static int sof_parse_tokens(struct snd_soc_component *scomp,
 		/* next array */
 		array = (struct snd_soc_tplg_vendor_array *)((u8 *)array
 			+ asize);
+
+		/* move to next target struct */
+		if (found == count) {
+			offset += object_size;
+			total += found;
+			found = 0;
+		}
```

Comment:
it should run until end of area and return -EINVAL

Reply:
@juimonen isn't there a possibility, that if `found` on one of iterations jumps from `< count` to `> count`, then the loop will continue until it runs out of space, but then it will return 0 - success?

Reply:
@lyakh the count is given to subfunctions and subfunctions ++ the found max count times... so found should never be more than count.(?)

Reply:
is it `found++` or `found += N`? It looks like `found` can be incremented by more than 1 on a single iteration. This is what I'm thinking about: say, iteration 0 `count` = 4, `found` = 0. You go to one of parsers and it returns 3. That's possible, right? Now 3 != 4, so you don't reset `found`. You go to the next iteration and again call the parser with `count` = 4, so, it can return anything `<= 4`, say, it returns 3 again. Now `found` = 6...

Reply:
@lyakh I guess the main point here is that you want to return error if this function doesn't find the amount of tokens it was asked to find? So what if I just add after the while loop:
if (total != count * sets)
    return -EINVAL;
?

It would then be error if we found anything else than the requested amount. The "found" should not matter as someone might spread the tokens into different array (which is strange also, but that would be the case you we're referring).

Reply:
@juimonen What is the expectation from each of the `sof_parse_*_tokens()` function? Are they expected to find exactly `count` elements? What if they find fewer of those? It *looks* like you continue to the next set, but then you expect to only find up to `count - found` elements in such a case, don't you? So maybe you should be using `count - found` as an argument to them? Then you would guarantee that `found` never exceeds `count`. But I really don't quite understand the *protocol* - how many elements of the expected properties are expected to be found in each set and what does it mean if fewer elements are found?

Reply:
@lyakh no, they might find more or less or exact amount as different token types might make up a "set".  so set might be like 2 uuids and 1 string. And array here is not same as "set". array is the alsa topology concept that packages a token. All tokens are in arrays, despite they belong to "artificial" composition called "set". So sets are like the "_tokens" structs on top of this file i.e. composition of multiple tokens we try to set for ipc.

So someone can put less tokens to topology or multiple overlapping tokens (so this doesn't make any sense, it is a error case for us). Before this PR this function was just blindly copying on top of previous found item. If you had less tokens than count, you would run until the end (and I think return success still, or actually this was "void" so you would not even know that).

Reply:
@plbossart @ranj063 as chatting today with @lyakh, I was really testing what happens 
with different "token cases", so FYI (and sorry for the long report): 

I was running basic sof-apl-pcm512x.tplg in up2, findings:

We search for following tokens, but don't care if they don't exist
(because they don't exist in this topology and it is successfully loaded):

pcm: dmac_tokens
volume_control: led_tokens
dai: stream_tokens
link: hda_tokens (which btw token array is empty so count = 0?)

So I can see that we have tokens which are _optional_, so we should
be able to search for those but not bail out.

In this sense sof_parse_token_sets is returning error only
in case:
- source token array size coming from user space is > 0
- dest data priv_size is enough to hold the token
- uuid type in source token array is supported

So currently sof_parse_token_(sets) doesn't care if the token is found or not,
just that parameters given to it are sane.

I tried the following error cases:

1) duplicates, more values than suppose to be there

tuples."short.pdm0" {
SOF_TKN_INTEL_DMIC_PDM_CTRL_ID          "0"
SOF_TKN_INTEL_DMIC_PDM_MIC_A_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_MIC_B_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_A       "0"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_B       "0"
SOF_TKN_INTEL_DMIC_PDM_CLK_EDGE "0"
SOF_TKN_INTEL_DMIC_PDM_SKEW             "0"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_A       "2"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_B       "2"
SOF_TKN_INTEL_DMIC_PDM_CLK_EDGE "2"
SOF_TKN_INTEL_DMIC_PDM_SKEW             "2"
}

In this case alsa_tplg is parsing the duplicates out and
the last set values remain, when converting from conf to tplg binary.
I could try to hack the binary but probably doesn't make sense...

*********************************

2) too few values

tuples."short.pdm0" {
SOF_TKN_INTEL_DMIC_PDM_CTRL_ID          "0"
SOF_TKN_INTEL_DMIC_PDM_MIC_A_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_MIC_B_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_A       "0"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_B       "0"
}

No error, we are just not setting the values in ipc struct,
so they are the default what was in ipc struct creation.

*********************************

3) 2 sets with 1 too few values

tuples."short.pdm0" {
SOF_TKN_INTEL_DMIC_PDM_CTRL_ID          "0"
SOF_TKN_INTEL_DMIC_PDM_MIC_A_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_MIC_B_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_A       "0"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_B       "0"
}

tuples."short.pdm1" {
SOF_TKN_INTEL_DMIC_PDM_CTRL_ID          "1"
SOF_TKN_INTEL_DMIC_PDM_MIC_A_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_MIC_B_Enable     "1"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_A       "0"
SOF_TKN_INTEL_DMIC_PDM_POLARITY_B       "0"
SOF_TKN_INTEL_DMIC_PDM_CLK_EDGE "0"
SOF_TKN_INTEL_DMIC_PDM_SKEW             "0"
}

because found is in first iteration 5 then 7
it is newer equal to count, so offset is not updated
and all values (even duplicates) are copied on top of
first set. 2nd set will remain in default values from
ipc struct creation.

*********************************

So questions:
1) are 2 and 3 errors as all ipc struct values in the set aren't found?
2) should we return error if not all are found and change all optional token
parsing in upper level to handle that, mandatory ones should bail out with error?

Reply:
Thanks @juimonen. That's a bit surprising. Reading parse_tokens documentation, it would seem that all the tokens are mandatory, but based on your analysis, with our existing topologies, we in practise have optional elements, so if we start reporting errors, we can expect many failures from existing topologies, right? I'd vote to make the parsing as strict as possible, given current topologies, but no more.

Reply:
@kv2019i yes, as this is not mostly my original code, I really can't trace the original author's idea about the parsing... Anyway as said, I can make the parsing more strict and handle the optional cases I found. But that would be probably new commit or perhaps a PR and not strictly related to this PR. And would need even more extensive testing in all possible platforms.

So should I try to fix that? I can see try to see what happens, this PR is not in panic to get merged...

Reply:
@juimonen At least not for this PR. Let's fix one issue at a time.